File: training_app_mobile/lib/models/exercise.dart
class Exercise {
  final String id;
  final String userId;
  final String exerciseTypeId;
  final String exerciseTypeName;
  final String? exerciseTypeCategory;
  final String? workoutId;
  final String? name;
  final String? notes;
  final Map<String, dynamic> metadata;
  final DateTime performedAt;
  final DateTime createdAt;
  final DateTime updatedAt;

  Exercise({
    required this.id,
    required this.userId,
    required this.exerciseTypeId,
    required this.exerciseTypeName,
    this.exerciseTypeCategory,
    this.workoutId,
    this.name,
    this.notes,
    required this.metadata,
    required this.performedAt,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Exercise.fromJson(Map<String, dynamic> json) {
    return Exercise(
      id: json['id'],
      userId: json['user_id'],
      exerciseTypeId: json['exercise_type_id'],
      exerciseTypeName: json['exercise_type_name'],
      exerciseTypeCategory: json['exercise_type_category'],
      workoutId: json['workout_id'],
      name: json['name'],
      notes: json['notes'],
      metadata: json['metadata'] ?? {},
      performedAt: DateTime.parse(json['performed_at']),
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'exercise_type_id': exerciseTypeId,
      'exercise_type_name': exerciseTypeName,
      'exercise_type_category': exerciseTypeCategory,
      'workout_id': workoutId,
      'name': name,
      'notes': notes,
      'metadata': metadata,
      'performed_at': performedAt.toIso8601String(),
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }

  String get displayName {
    return name ?? exerciseTypeName;
  }
}

class CreateExerciseRequest {
  final String exerciseTypeId;
  final String? workoutId;
  final String? name;
  final String? notes;
  final Map<String, dynamic>? metadata;
  final DateTime? performedAt;

  CreateExerciseRequest({
    required this.exerciseTypeId,
    this.workoutId,
    this.name,
    this.notes,
    this.metadata,
    this.performedAt,
  });

  Map<String, dynamic> toJson() {
    final json = <String, dynamic>{
      'exercise_type_id': exerciseTypeId,
    };

    if (workoutId != null) json['workout_id'] = workoutId;
    if (name != null) json['name'] = name;
    if (notes != null) json['notes'] = notes;
    if (metadata != null) json['metadata'] = metadata;
    if (performedAt != null) json['performed_at'] = performedAt!.toIso8601String();

    return json;
  }
}

class UpdateExerciseRequest {
  final String? exerciseTypeId;
  final String? workoutId;
  final String? name;
  final String? notes;
  final Map<String, dynamic>? metadata;
  final DateTime? performedAt;

  UpdateExerciseRequest({
    this.exerciseTypeId,
    this.workoutId,
    this.name,
    this.notes,
    this.metadata,
    this.performedAt,
  });

  Map<String, dynamic> toJson() {
    final json = <String, dynamic>{};

    if (exerciseTypeId != null) json['exercise_type_id'] = exerciseTypeId;
    if (workoutId != null) json['workout_id'] = workoutId;
    if (name != null) json['name'] = name;
    if (notes != null) json['notes'] = notes;
    if (metadata != null) json['metadata'] = metadata;
    if (performedAt != null) json['performed_at'] = performedAt!.toIso8601String();

    return json;
  }
}

File: backend/src/models/exercise.rs
use serde::{Deserialize, Serialize};
use time::OffsetDateTime;
use uuid::Uuid;
use validator::Validate;

#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct Exercise {
    pub id: Uuid,
    pub user_id: Uuid,
    pub exercise_type_id: Uuid,
    pub workout_id: Option<Uuid>,
    pub name: Option<String>,
    pub notes: Option<String>,
    pub metadata: serde_json::Value,
    #[serde(with = "time::serde::rfc3339")]
    pub performed_at: OffsetDateTime,
    #[serde(with = "time::serde::rfc3339")]
    pub created_at: OffsetDateTime,
    #[serde(with = "time::serde::rfc3339")]
    pub updated_at: OffsetDateTime,
    #[serde(with = "time::serde::rfc3339::option")]
    pub deleted_at: Option<OffsetDateTime>,
}

#[derive(Debug, Serialize, Deserialize, Validate)]
pub struct CreateExerciseRequest {
    pub exercise_type_id: Uuid,
    pub workout_id: Option<Uuid>,
    #[validate(length(min = 1, max = 255, message = "Name must be between 1 and 255 characters"))]
    pub name: Option<String>,
    #[validate(length(max = 2000, message = "Notes cannot exceed 2000 characters"))]
    pub notes: Option<String>,
    pub metadata: Option<serde_json::Value>,
    #[serde(default, with = "time::serde::rfc3339::option")]
    pub performed_at: Option<OffsetDateTime>,
}

#[derive(Debug, Serialize, Deserialize, Validate)]
pub struct UpdateExerciseRequest {
    pub exercise_type_id: Option<Uuid>,
    pub workout_id: Option<Uuid>,
    #[validate(length(min = 1, max = 255, message = "Name must be between 1 and 255 characters"))]
    pub name: Option<String>,
    #[validate(length(max = 2000, message = "Notes cannot exceed 2000 characters"))]
    pub notes: Option<String>,
    pub metadata: Option<serde_json::Value>,
    #[serde(default, with = "time::serde::rfc3339::option")]
    pub performed_at: Option<OffsetDateTime>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ExerciseResponse {
    pub id: Uuid,
    pub user_id: Uuid,
    pub exercise_type_id: Uuid,
    pub exercise_type_name: String,
    pub exercise_type_category: Option<String>,
    pub workout_id: Option<Uuid>,
    pub name: Option<String>,
    pub notes: Option<String>,
    pub metadata: serde_json::Value,
    #[serde(with = "time::serde::rfc3339")]
    pub performed_at: OffsetDateTime,
    #[serde(with = "time::serde::rfc3339")]
    pub created_at: OffsetDateTime,
    #[serde(with = "time::serde::rfc3339")]
    pub updated_at: OffsetDateTime,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ExerciseWithType {
    pub id: Uuid,
    pub user_id: Uuid,
    pub exercise_type_id: Uuid,
    pub exercise_type_name: String,
    pub exercise_type_category: Option<String>,
    pub exercise_type_metadata_schema: serde_json::Value,
    pub workout_id: Option<Uuid>,
    pub name: Option<String>,
    pub notes: Option<String>,
    pub metadata: serde_json::Value,
    #[serde(with = "time::serde::rfc3339")]
    pub performed_at: OffsetDateTime,
    #[serde(with = "time::serde::rfc3339")]
    pub created_at: OffsetDateTime,
    #[serde(with = "time::serde::rfc3339")]
    pub updated_at: OffsetDateTime,
}

impl From<ExerciseWithType> for ExerciseResponse {
    fn from(exercise: ExerciseWithType) -> Self {
        Self {
            id: exercise.id,
            user_id: exercise.user_id,
            exercise_type_id: exercise.exercise_type_id,
            exercise_type_name: exercise.exercise_type_name,
            exercise_type_category: exercise.exercise_type_category,
            workout_id: exercise.workout_id,
            name: exercise.name,
            notes: exercise.notes,
            metadata: exercise.metadata,
            performed_at: exercise.performed_at,
            created_at: exercise.created_at,
            updated_at: exercise.updated_at,
        }
    }
}